File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionDigest.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.functions.common.*;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Base64;
import java.util.HexFormat;
import java.util.Map;

public class TransformerFunctionDigest extends TransformerFunction {
    public TransformerFunctionDigest() {
        super(FunctionDescription.of(
            Map.of(
            "algorithm", ArgumentType.of(ArgType.Enum).position(0).defaultEnum("SHA-1"),
            "format", ArgumentType.of(ArgType.Enum).position(1).defaultEnum("BASE64")
            )
        ));
    }
    @Override
    public CompletionStage<Object> apply(FunctionContext context) {
        var str = context.getString(null);
        if (str == null) {
            return null;
        }
        var algorithm = context.getEnum("algorithm");
        if (algorithm.equals("JAVA")) {
            return str.hashCode();
        }
        try {
            var digest = MessageDigest.getInstance(algorithm).digest(str.getBytes());
            return switch (context.getEnum("format")) {
                case "BASE64" -> Base64.getEncoder().encodeToString(digest);
                case "BASE64URL" -> Base64.getUrlEncoder().encodeToString(digest);
                default -> HexFormat.of().formatHex(digest);
            };
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }
}

File: javascript/json-transform/src/functions/TransformerFunctionDigest.ts
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";
import TextEncoding from "./common/TextEncoding";
import Base64 from "./utils/Base64";
import md5 from "./utils/md5";

const globalCrypto = typeof global.crypto !== "undefined" ? global.crypto : require("crypto");

function hashCode(input: string) {
  let h: Int32Array | null = new Int32Array(1);
  for (let i = 0; i < input.length; i++) {
    h[0] = 31 * h[0] + (input.charCodeAt(i) & 0xff);
  }
  const result = h[0];
  h = null; // free buffer
  return result;
}
function formatHex(a: ArrayBuffer) {
  return [...new Uint8Array(a)].map(x => x.toString(16).padStart(2, "0")).join("");
}

class TransformerFunctionDigest extends TransformerFunction {
  constructor() {
    super({
      arguments: {
        algorithm: { type: ArgType.Enum, position: 0, defaultEnum: "SHA-1" },
        format: { type: ArgType.Enum, position: 1, defaultEnum: "BASE64" },
      },
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const str = await context.getString(null);
    if (str == null) {
      return null;
    }
    const algorithm = await context.getEnum("algorithm");
    if (algorithm === "JAVA") {
      return hashCode(str);
    }
    const digest = !algorithm
      ? new ArrayBuffer(0)
      : algorithm === "MD5"
        ? md5(str)
        : await globalCrypto.subtle.digest({ name: algorithm }, TextEncoding.encode(str, "ISO-8859-1"));
    switch (await context.getEnum("format")) {
      case "BASE64":
        return Base64.encode(new Uint8Array(digest), "basic");
      case "BASE64URL":
        return Base64.encode(new Uint8Array(digest), "url");
      default:
        return formatHex(digest);
    }
  }
}

export default TransformerFunctionDigest;
